File: AirshipKit/AirshipKit/common/UADelay.m
/* Copyright Urban Airship and Contributors */

#import "UADelay+Internal.h"

@interface UADelay()
@property (nonatomic, assign) NSTimeInterval seconds;
@property (nonatomic, strong) dispatch_semaphore_t semaphore;    // GCD objects use ARC
@end

@implementation UADelay

- (instancetype)initWithDelayInSeconds:(NSTimeInterval)seconds {
    self = [super init];
    if (self) {
        self.seconds = seconds;
        self.semaphore = dispatch_semaphore_create(0);
    }
    return self;
}

+ (instancetype)delayWithSeconds:(NSTimeInterval)seconds {
    return [[UADelay alloc] initWithDelayInSeconds:seconds];
}

- (void)cancel {
    dispatch_semaphore_signal(self.semaphore);
}

- (void)start {
    //dispatch time is calculated as nanoseconds delta offset
    dispatch_semaphore_wait(self.semaphore, dispatch_time(DISPATCH_TIME_NOW, (self.seconds * NSEC_PER_SEC)));
}

@end

File: Airship/AirshipCore/Source/UADelay.swift
/* Copyright Airship and Contributors */

/// Delay protocol.
/// For internal use only. :nodoc:
@objc(UADelayProtocol)
public protocol DelayProtocol {
    @objc
    func cancel()

    @objc
    func start()
}

/**
 * @note For internal use only. :nodoc:
 */
@objc
public class UADelay : NSObject, DelayProtocol {
    private let seconds: TimeInterval
    private let semaphore: DispatchSemaphore

    @objc
    public init(_ seconds: TimeInterval) {
        self.seconds = seconds
        semaphore = DispatchSemaphore(value: 0)
    }

    @objc
    public func cancel() {
        semaphore.signal()
    }

    @objc
    public func start() {
        _ = semaphore.wait(timeout: DispatchTime.now() + seconds)
    }
}
